### PR TITLE
Fix const range reduction when a float parameter is used

### LINKED ISSUE #59545: GDScript 2.0: Zero vector when using `const` instead of `var` in a specific scenario
### Godot version

4.0.alpha5.official (d7d528c15)

### System information

Windows 10 Pro 21H1

### Issue description

In a very specific scenario, using `const` instead of `var` in GDScript changes the outcome, resulting in a created vector incorrectly being a zero vector. Using `const` instead of `var` should only affect the mutability of the variable, not the result of the operation.

### Steps to reproduce

The following code should work in any attached script. I have tested it in the `_ready()` function of two unrelated scene objects.

## Incorrect, with const
```
const COUNT = 10
const DIV = 5
for a in range(ceil(COUNT / DIV)):
	var vec = Vector3i(a, a, a)
	print(vec, " ", a, " ", COUNT, " ", DIV)
```
incorrectly prints
```
(0, 0, 0) 0 10 5
(0, 0, 0) 1 10 5
```
## Correct, with var
```
const COUNT = 10
var DIV = 5
for a in range(ceil(COUNT / DIV)):
	var vec = Vector3i(a, a, a)
	print(vec, " ", a, " ", COUNT, " ", DIV)
```
correctly prints
```
(0, 0, 0) 0 10 5
(1, 1, 1) 1 10 5
```

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by cdemirer**
Fixes #59545

### CODE REVIEW

File: modules/gdscript/gdscript_analyzer.cpp [hunk 1]
```diff
@@ -1362,7 +1362,13 @@ void GDScriptAnalyzer::resolve_for(GDScriptParser::ForNode *p_for) {
 					if (all_is_constant) {
 						switch (args.size()) {
 							case 1:
-								reduced = args[0];
+								if (args[0].get_type() == Variant::FLOAT) {
+									Callable::CallError temp;
+									const Variant *construct_arg = &args[0];
+									Variant::construct(Variant::INT, reduced, &construct_arg, 1, temp);
```
Comment (cdemirer) [inline modules/gdscript/gdscript_analyzer.cpp:1368]: There should probably be error checking here since the float could be something like `INF`

Comment (vnen) [reply modules/gdscript/gdscript_analyzer.cpp:1368]: As discussed in the meeting, having this check for INF and NaN would be nice.

**Comment by vnen** [comment]
Looks like a good fix.

File: modules/gdscript/gdscript_analyzer.cpp (hunk 1 above)
Comment (cdemirer) [reply modules/gdscript/gdscript_analyzer.cpp:1368]: Just found out that there is no check for normal `range` yet.
```gdscript
var x = range(INF) # INF gets converted to int (becomes INT_MIN) and used directly. Returns `[]`
```
Also if it should be checked, should the other 2 parameters be checked as well?
During runtime for non constant `range`?

I started to think maybe the checks can be added with another PR... (as the Variant::construct here never "fails", it's as safe as before)

**Comment by dalexeev** [comment]
Looks like the related issue has already been fixed by #68125 and #71279.

**Comment by YuriSizov** [comment]
Closing as superseded. Thanks for your contribution nonetheless!

**Pull request closed without merging**